### ISSUE TITLE
Support expression when search a collection

### ISSUE THREAD

**Issue by DragonDriver** [milvus-io/milvus#5028]
**Please state your issue using the following template and, most importantly, in English.**

**Is your feature request related to a problem? Please describe.**

I want milvus to support expression to filter attribute when search a collection.

**Describe the solution you'd like**

We should change `SearchRequest` in `milvus.proto` by adding serveral field which is related to expression.

```proto
message SearchRequest {
  common.MsgBase base = 1;
  string db_name = 2;
  string collection_name = 3;
  repeated string partition_names = 4;
  string dsl = 5;
  // serialized `PlaceholderGroup`
  bytes placeholder_group = 6;
  bool is_expr = 7;
  string expr = 8;
  string params = 9;    // serialized json string
  string anns_field = 10;
}
```

To keep compatible with dsl version, we should retain the `dsl` field. `is_expr` indicates whether the search request use expression, when `is_expr` is set to `true`, then `expr`, `params`and `anns_field` are enable. `expr` field is often used to filter attribute. For example, a collection has two field, `year` and `embedding`, whose type are int64 and float vector, we can use `year > 1998` to filter records whose year is less than 1998. `params` stores the search parameters, such as `topk`, `nprobe`, etc. `anns_field` tells milvus which field is used to search.

To support this feature, first, we need `segcore` support expression, then we need a patch in `proxy`. When `proxy` receive search requests from clients, `proxy` will check `is_expr` field in request, when `is_expr` is set to true, `proxy` will translate `expr`, `params`and `anns_field` to `dsl` field.

**Describe alternatives you've considered**

**Additional context**
Add any other context, code examples, or references to existing implementations about the feature request here.

**Comment by DragonDriver** [comment]
The part of `segcore` will be implemented by @FluorineDog

**Comment by FluorineDog** [comment]
```
 string params = 9;    // serialized json string
 string anns_field = 10;
```
params -> search_params
anns_field -> anns_field_name

**Comment by czs007** [comment]
I suggest to modify the proto to be like this：

message SearchRequest {
  common.MsgBase base = 1;
  string db_name = 2;
  string collection_name = 3;
  repeated string partition_names = 4;
  string dsl = 5;
  bytes placeholder_group = 6;
  common.DslType dsl_type = 7;
  repeated common.KeyValuePair search_params = 8;
}

common.DslType is defined as follows：

enum DslType {
    Dsl = 0
    BoolExprV1 = 1
}

When filed `dsl_type`  is DslType.Dsl , you can specify a json string compatible with the definition of ES's dsl to field `dsl`.
When filed `dsl_type` is DslType.BoolExprV1, you can speficy a string similar to the boolean expressions supported by pandas.DataFrame.query interface to the field `dsl`.

In addition, pack all the other params to field 'search_params'.

In common.KeyValuePair, both Key and Value are type of string. What needs to be pointed out is that the Value needs to be generated using a unified object serialization library, such as json. You cannot use the __str__ method of the Python built-in object.

**Comment by DragonDriver** [comment]
> I suggest to modify the proto to be like this：
> 
> message SearchRequest {
> common.MsgBase base = 1;
> string db_name = 2;
> string collection_name = 3;
> repeated string partition_names = 4;
> string dsl = 5;
> bytes placeholder_group = 6;
> common.DslType dsl_type = 7;
> repeated common.KeyValuePair search_params = 8;
> }
> 
> common.DslType is defined as follows：
> 
> enum DslType {
> Dsl = 0
> BoolExprV1 = 1
> }
> 
> When filed `dsl_type` is DslType.Dsl , you can specify a json string compatible with the definition of ES's dsl to field `dsl`.
> When filed `dsl_type` is DslType.BoolExprV1, you can speficy a string similar to the boolean expressions supported by pandas.DataFrame.query interface to the field `dsl`.
> 
> In addition, pack all the other params to field 'search_params'.
> 
> In common.KeyValuePair, both Key and Value are type of string. What needs to be pointed out is that the Value needs to be generated using a unified object serialization library, such as json. You cannot use the **str** method of the Python built-in object.

OK, i think it's very reasonable to use `dsl_type`. By the way, `anns_field` or `anns_field_name` is also stored in `search_params`?

**Comment by czs007** [comment]
> > I suggest to modify the proto to be like this：
> > message SearchRequest {
> > common.MsgBase base = 1;
> > string db_name = 2;
> > string collection_name = 3;
> > repeated string partition_names = 4;
> > string dsl = 5;
> > bytes placeholder_group = 6;
> > common.DslType dsl_type = 7;
> > repeated common.KeyValuePair search_params = 8;
> > }
> > common.DslType is defined as follows：
> > enum DslType {
> > Dsl = 0
> > BoolExprV1 = 1
> > }
> > When filed `dsl_type` is DslType.Dsl , you can specify a json string compatible with the definition of ES's dsl to field `dsl`.
> > When filed `dsl_type` is DslType.BoolExprV1, you can speficy a string similar to the boolean expressions supported by pandas.DataFrame.query interface to the field `dsl`.
> > In addition, pack all the other params to field 'search_params'.
> > In common.KeyValuePair, both Key and Value are type of string. What needs to be pointed out is that the Value needs to be generated using a unified object serialization library, such as json. You cannot use the **str** method of the Python built-in object.
> 
> OK, i think it's very reasonable to use `dsl_type`. By the way, `anns_field` or `anns_field_name` is also stored in `search_params`?

Yeah,  anns_field only needs to be specified when the  `dsl_type` is BoolExprV1 . Different DslType has different matching parameter sets.

**Comment by FluorineDog** [comment]
Change Recommendation:

```protobuf
message SearchRequest {

  common.MsgBase base = 1;
  string db_name = 2;
  string collection_name = 3;
  repeated string partition_names = 4;
  string dsl = 5;
  bytes placeholder_group = 6;
  common.DslType dsl_type = 7;
  // example: [<"topk", "10">, <"metric_type", "L2">, <"search_params", """{"nprobe": 10}""">]
  map<string, string> extra_params = 8;
}

// common.DslType is defined as follows：
enum DslType {
  Dsl = 0
  BoolExprV1 = 1
}

```

**Comment by FluorineDog** [comment]
resolved by #5096 #5088
